### DISCUSSION TITLE
Question: Why Homebrew requires full rebuild to enable optional dependencies?

### DISCUSSION THREAD

**Discussion by equal-l2** [Homebrew/discussions#1673] [Everyday usage] [answered]
This is NOT a complaint nor a feature request. I just want to how things work.
If you know a more appropriate place to ask this kind of question, please tell me, as I'm not so confident about if here is the right place to ask.

---

With package managers on Linux (such as *apt* or *pacman*), I can enable an optional dependency just by installing it. 
However, with Homebrew, rebuilding a formula from the source is required to enable an optional dependency.

Is that because of a difference between Linux and macOS (API, the format of libraries, or else)?
Or is that because of something specific with Homebrew (lack of resource, design decision)?

**Comment by SMillerDev** [comment, marked as answer]
>  Or is that because of something specific with Homebrew (lack of resource, design decision)?

It's mostly a difference of goals. We do have formula that provide optional dependencies without recompiling. A lot of libraries in homebrew provide `.pc` files so you could see `pkg-config` as an optional dependency for them. We just don't advertise those because library users will know when to use `pkg-config`.

The optional dependencies as defined in https://docs.brew.sh/Formula-Cookbook#specifying-other-formulae-as-dependencies are different though. They act more like "compile modes" and since homebrew usually can't compile software with all options enabled (think python 2 or 3) this requires a recompile. For Linux these would usually be a completely different package but homebrew took another approach.
